src/dhtproto/client/legacy/internal/helper/Handshake.d

```diff
@@ -0,0 +1,211 @@
+        }
+    }
+
+
+    /***************************************************************************
+
+        Logger for the handshake
+
+    ***************************************************************************/
+
+    private Logger log;
+
+
+    /***************************************************************************
+
+        Instance of task that suspends until all nodes are connected
+
+    ***************************************************************************/
+
+    private AllDhtNodesConnected all_nodes_connected;
+
+
+    /***************************************************************************
+
+        Instance of task that suspends until at least one node is connected
+
+    ***************************************************************************/
+
+    private OneDhtNodeConnected one_node_connected;
+
+
+    /***************************************************************************
+
+        Instance of class that launches the actual handshake, retrying when
+        it fails until all nodes are connected
+
+    ***************************************************************************/
+
+    private RetryHandshake retry_handshake;
+
+
+    /***************************************************************************
+
+        Constructor
+
+        Params:
+            epoll = epoll select dispatcher via which to schedule handshake
+            dht = DHT client instance via which to perform the handshake
+            wait_time = time to wait (in seconds) before retrying failed
+                handshake attempts
+
+    ***************************************************************************/
+
+    public this (EpollSelectDispatcher epoll, DhtClient dht, ulong wait_time)
+    in
+    {
+        assert(dht !is null);
+    }
+    body
+    {
+        this.log = Log.lookup("dhtproto.client.legacy.internal.helper.Handshake");
+
+        this.all_nodes_connected = new AllDhtNodesConnected;
+        this.one_node_connected = new OneDhtNodeConnected;
+
+        this.retry_handshake =
+            new RetryHandshake(epoll, dht, wait_time,
+                    &this.all_nodes_connected.dg, &this.one_node_connected.dg);
+    }
+
+
+    /***************************************************************************
+
+        Returns:
+            instance of task that suspends until all nodes have been connected
+
+    ***************************************************************************/
+
+    public AllDhtNodesConnected allNodesConnected ()
+    {
+        return this.all_nodes_connected;
+    }
+
+
+    /***************************************************************************
+
+        Returns:
+            instance of task that suspends until at least one node is connected
+
+    ***************************************************************************/
+
+    public OneDhtNodeConnected oneNodeConnected ()
+    {
+        return this.one_node_connected;
+    }
+}
```

Comment:
Please add a usage example (like what you included in the commit message).

Reply:
Yup, will do so.  Just wanted to get the design itself tied down/approved first.

How feasible is it to really _test_ something like this?  I don't have much experience of swarm/proto unittesting and integration tests.

Reply:
Still missing usage example